### ISSUE TITLE
The hint and solution articles for curriculum challenges need some guidelines

### ISSUE THREAD

**Issue by RandellDawson** [freeCodeCamp/freeCodeCamp#35453]
The purpose of this issue is to brainstorm some ways we could simplify the structure of the Guide articles in the `certifications` folder and recommend some basic guidelines which should be followed, so contributors understand what qualifies as an "acceptable" change/addition to these specific articles.  These articles are what a camper sees when clicking on the `Get a Hint` button on each challenge, so I feel it is important the camper's experience when viewing the articles is consistent across all of them.

Below are what I consider the minimum requirements for such a Guide article:

* Two headers (Hints and Solutions)
* For Hints section, there can be zero up to 3-5 hints maximum
* Move away from classifying solutions using subjective terms of Basic, Intermediate, and Advanced.  Instead, the solutions would display the solution's Time Complexity and Space Complexity.  As far as order is concerned, we can sort them in terms of Time Complexity
* Where multiple solutions can exist (i.e. algorithm challenges), only allow a max of 2-3 solutions within the same Time Complexity/Space Complexity classification.  This will prevent the article from showing the same basic algorithm but only differing in syntax or some logic order (i.e using higher order functions instead of for or while loops). 
* There should be as few images as possible unless they are critical to a hint or solution explanation.  Currently, the current design has too many icons and distracts from the content.
* All solutions should be full working solutions and include all Seed Code, so that a camper could simply copy/paste entire solution into editor and see it pass all tests. A nice to have would be to put each solution in a Disclosure Summary element like below so the solutions are hidden upon entering the page in case a user just wants a hint and not see the full solution right away.  For example, the following syntax:
````html

<details><summary>Example #1 (Click to Unhide)</summary>

```js
function myFunc() {
  console.log('Hello World!');
}
```

</details>
````
would produce the following effect:

<details><summary>Example &#35;1 (Click to Unhide)</summary>

```js
function myFunc() {
  console.log('Hello World!');
}
```

</details>
<br>

* Remove all links to replt.it, Codepen, jsFiddle, etc. as the camper can always copy/paste the solution into the FCC code editor to see the code works.
* Make code explanations optional.  Maybe a valid solution is enough and then if others want to make improvements and add code explanations in other PRs, that would be fine too. 
---
I think the above points gets us started.  These are just my opinions based on reviewing hundreds of PRs concerning changes to the challenge solutions in this hint section.  Would like to hear from mods and contributors on this issue.

**Comment by thecodingaviator** [comment]
All the points seem fine, except I think it wouldn't be very wise to put a heading for Hints if we're going to have zero

**Comment by RandellDawson** [comment]
@thecodingaviator What I am wondering is if we can leverage some of the logic which is currently used in the curriculum challenges.  The main header sections would only render if there was a `section` with `id='hints'` or `id='solutions``.  Otherwise, they would not render at all. That way if one or both of the sections did not have hints or solutions, those sections would not render at all.  If no sections were detected, then some kind of default stub would render.  

Below is just an example of how one would possibly write the hints and solutions and then the display logic would render.  We might be able to use yaml like we do in the Tests section of each challenge instead of having to write out tags like `<div class="hint">...</div>`.  However it is implemented, the desired effect is to display the page something like I show at the bottom of this comment.  I am sure we can add some styling to make it more visually appealing, but hopefully the concept is my main point.  I like the idea that hints and solutions are hidden by default, so the camper can select which ones they want to view instead of just seeing them immediately.
````

<section id='hints'>
<div class="hint">
Hint goes here
</div>
<div class="hint">
hint goes here
</div>
</section>

<section id='solutions'>
<div class="solution">

```js
function myFunc() {
  console.log('Hello World!');
}
```
#### Code Explanation
code explanation goes here
</div>

<div class="solution">
```js
const myFunc = () => console.log('Hello World!');
```

</div>
</section>
````

The above would automatically display as:

## Hints

<details><summary>Hint &#35;1 (Click to Unhide)</summary>
hint goes here
</details>

<br>
<details><summary>Hint &#35;2 (Click to Unhide)</summary>
hint goes here
</details>

## Solutions

<details><summary>Solution &#35;1 (Click to Unhide)</summary>

```js
function myFunc() {
  console.log('Hello World!');
}
```
#### Code Explanation
code notes or explanation here
</details>
<br>
<details><summary>Solution &#35;2 (Click to Unhide)</summary>

```js
const myFunc = () => console.log('Hello World!');
```
</details>

**Comment by Bouncey** [comment]
This sounds like a solid strategy with regards to curriculum articles.

> What I am wondering is if we can leverage some of the logic which is currently used in the curriculum challenges.

The way the challenge markdown files are consumed by `gatsby` is completely different to guide markdown files and would require a lot of work to implement. For challenges, we have the `challenge-md-parser` which converts the markdown to a challenge object, it was necessary to 'roll our own'. 

The way to not render a Hints header in a guide article it to not have it in the markdown in the first place. You could comment it out with html comments if you wanted to preserve a template as such.

**Comment by thecodingaviator** [comment]
> The main header sections would only render if there was a `section` with `id='hints'` or `id='solutions``. Otherwise, they would not render at all. That way if one or both of the sections did not have hints or solutions, those sections would not render at all. If no sections were detected, then some kind of default stub would render.

I agree with what Stuart says above, it does definitely seem like a great strategy to go ahead with this, if we can pull it off, that is, we'd probably also need to go through 90% if not all files and make required changes

**Comment by moT01** [comment]
This sounds like a good idea, since the current hints have very little consistency across the guide. I pretty much like all these suggestions.

If we go with the collapsible sections, I would have the hints always displayed and not collapsible, and keep the solutions collapsed on load.

Also, maybe take into account that there is different types of challenges... classic, backend, and projects - so perhaps we want the project hints to have a slightly different structure or something from the others.

**Comment by MoonCWang** [comment]
Agree with all points especially doing away icons.  Some thoughts.

- Should current guide contributors wait for standardization before modifying their contributions?
- I feel hints and complex time order solutions don't have to be hidden.  Constant and linear time order should be.

**Comment by thecodingaviator** [comment]
**Point for discussion 1** : Should we add solutions to guide articles including methods that have *not been taught* in the curriculum yet?

**Point for discussion 2** : Should the guide articles be kept simplified? An example of what I mean is given in #35592 . Two methods which were written differently have been put in one, this might cause confusion to someone not familiar with with how the syntax works

**Comment by ojeytonwilliams** [comment]
Point 1: I think it's fine to include a solution using methods that have not been taught at that point.  However, they should only be added if there is already is a solution using methods they have been exposed to. Also, there needs to be a link introducing the concept.  This also serves as a cue to the reader that they are not expected to already understand what's being linked to.

It's important to keep in mind that people looking for hints are initially going to want things they can easily understand.  They may be motivated to subsequently learn smarter ways to do things, but not initially.

Point 2: Simple as in like in that PR?  Sure.  It's conciser than the original and at least as clear.  It's also in the Front End section, so it's fair to assume that the reader has completed the ES6 lessons before starting this challenge.

**Comment by moT01** [comment]
So, we've had a bunch of things we agree on here - I think someone should come up with a first draft of something to put in the guide somewhere, and we can fine tune it from there. And after we come up with a finalized version, we can probably make an issue for changing the current guide articles to conform to it.

**Comment by RandellDawson** [comment]
I will take a stab at it.  I think I know a way to make the majority happy.

**Comment by raisedadead** [comment]
Lets use this thread to discuss the direction for the template on the forum.

**Comment by RandellDawson** [comment]
This can be closed now, due to the documentation being updated to include the template to be used for the hint/solution forum topics.
